src/io/fileops.c

```diff
@@ -202,48 +215,47 @@ MVMint64 MVM_file_exists(MVMThreadContext *tc, MVMString *f, MVMint32 use_lstat)
 #ifdef _WIN32
 #define FILE_IS(name, rwx) \
     MVMint64 MVM_file_is ## name (MVMThreadContext *tc, MVMString *filename, MVMint32 use_lstat) { \
-        if (!MVM_file_exists(tc, filename, use_lstat)) \
+        uv_stat_t statbuf; \
+        if (file_info_with_error(tc, &statbuf, filename, use_lstat) < 0) { \
             return 0; \
+        } \
         else { \
-            uv_stat_t statbuf = file_info(tc, filename, use_lstat); \
             MVMint64 r = (statbuf.st_mode & S_I ## rwx ); \
             return r ? 1 : 0; \
         } \
     }
 FILE_IS(readable, READ)
 FILE_IS(writable, WRITE)
 MVMint64 MVM_file_isexecutable(MVMThreadContext *tc, MVMString *filename, MVMint32 use_lstat) {
-    if (!MVM_file_exists(tc, filename, use_lstat))
+    MVMint64 r = 0;
+    uv_stat_t statbuf;
+    if (file_info_with_error(tc, &statbuf, filename, use_lstat) < 0)
         return 0;
+    else if ((statbuf.st_mode & S_IFMT) == S_IFDIR)
+        return 1;
     else {
-        MVMint64 r = 0;
-        uv_stat_t statbuf = file_info(tc, filename, use_lstat);
-        if ((statbuf.st_mode & S_IFMT) == S_IFDIR)
-            return 1;
-        else {
-            /* true if fileext is in PATHEXT=.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC */
-            MVMString *dot = MVM_string_ascii_decode_nt(tc, tc->instance->VMString, ".");
-            MVMROOT(tc, dot, {
-                MVMint64 n = MVM_string_index_from_end(tc, filename, dot, 0);
-                if (n >= 0) {
-                    MVMString *fileext = MVM_string_substring(tc, filename, n, -1);
-                    char *ext  = MVM_string_utf8_c8_encode_C_string(tc, fileext);
-                    char *pext = getenv("PATHEXT");
-                    int plen   = strlen(pext);
-                    int i;
-                    for (i = 0; i < plen; i++) {
-                        if (0 == stricmp(ext, pext++)) {
-                             r = 1;
-                             break;
-                        }
+        /* true if fileext is in PATHEXT=.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC */
+        MVMString *dot = MVM_string_ascii_decode_nt(tc, tc->instance->VMString, ".");
```

Comment:
As a minor optimization, this could be one of the strings that we create at startup and cache on the `MVMInstance`.